File: smoke_tests/lua/ssrf.lua
-- Server-Side Request Forgery (SSRF) vulnerabilities in Lua

local http = require("socket.http")
local ltn12 = require("ltn12")

-- Test 1: HTTP request with user-controlled URL
function fetch_url(url)
    -- VULNERABLE: SSRF via user-controlled URL
    local response = {}
    http.request{
        url = url,
        sink = ltn12.sink.table(response)
    }
    return table.concat(response)
end

-- Test 2: OpenResty http client SSRF
function resty_http_fetch(url)
    local httpc = require("resty.http").new()
    -- VULNERABLE: User-controlled URL
    local res = httpc:request_uri(url)
    return res.body
end

-- Test 3: URL concatenation
function fetch_api(endpoint)
    -- VULNERABLE: Endpoint can be manipulated
    local url = "http://internal-api.local/" .. endpoint
    return http.request(url)
end

-- Test 4: ngx.location.capture SSRF
function proxy_request()
    local args = ngx.req.get_uri_args()
    local target = args.url
    -- VULNERABLE: Internal redirect to user URL
    local res = ngx.location.capture("/proxy", {
        vars = { target_url = target }
    })
    ngx.say(res.body)
end

-- Test 5: Socket connection to user host
function connect_to_host(host, port)
    local socket = require("socket")
    local tcp = socket.tcp()
    -- VULNERABLE: Connecting to arbitrary host
    tcp:connect(host, port)
    return tcp
end

-- Test 6: OpenResty socket SSRF
function resty_socket_connect()
    local args = ngx.req.get_uri_args()
    local host = args.host
    local port = args.port or 80
    local sock = ngx.socket.tcp()
    -- VULNERABLE: User-controlled socket connection
    sock:connect(host, tonumber(port))
end

-- Test 7: DNS rebinding potential
function fetch_with_host(host, path)
    -- VULNERABLE: Host header manipulation
    local url = "http://" .. host .. path
    return http.request(url)
end

-- Test 8: URL scheme manipulation
function fetch_resource(resource)
    -- VULNERABLE: Can use file:// or other schemes
    local url = resource
    return http.request(url)
end

-- Test 9: Redirect following
function fetch_with_redirect(url)
    local httpc = require("resty.http").new()
    -- VULNERABLE: May follow redirects to internal services
    local res = httpc:request_uri(url, {
        method = "GET",
        ssl_verify = false
    })
    return res.body
end

-- Test 10: Webhook URL
function send_webhook(webhook_url, data)
    local httpc = require("resty.http").new()
    -- VULNERABLE: User-controlled webhook destination
    httpc:request_uri(webhook_url, {
        method = "POST",
        body = data
    })
end

-- Test 11: Image/file URL fetch
function fetch_image(image_url)
    -- VULNERABLE: Fetching from user URL
    local response = {}
    http.request{
        url = image_url,
        sink = ltn12.sink.table(response)
    }
    return table.concat(response)
end

-- Test 12: XML with external entity
function process_xml_url(xml_url)
    -- VULNERABLE: Fetching XML from user URL
    local content = fetch_url(xml_url)
    -- Parse XML...
    return content
end

-- Test 13: Database connection string
function connect_db(connection_string)
    -- VULNERABLE: User-controlled connection destination
    local pgmoon = require("pgmoon")
    local pg = pgmoon.new(parse_connection_string(connection_string))
    pg:connect()
    return pg
end

function parse_connection_string(str)
    -- Parse and return connection params
    return {host = "localhost", database = "test"}
end

File: smoke_tests/dart/ssrf.dart
// SSRF (Server-Side Request Forgery) vulnerabilities in Dart

import 'dart:io';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:dio/dio.dart';

// Test 1: HttpClient with user URL
Future<String> vulnerableHttpClient(String userUrl) async {
  // VULNERABLE: SSRF via user-controlled URL
  var client = HttpClient();
  var request = await client.getUrl(Uri.parse(userUrl));
  var response = await request.close();
  return await response.transform(utf8.decoder).join();
}

// Test 2: http package with user URL
Future<http.Response> vulnerableHttpGet(String url) async {
  // VULNERABLE: SSRF via http.get
  return await http.get(Uri.parse(url));
}

// Test 3: http.post with user URL
Future<http.Response> vulnerableHttpPost(String url, String body) async {
  // VULNERABLE: SSRF via http.post
  return await http.post(Uri.parse(url), body: body);
}

// Test 4: Dio with user URL
Future<Response> vulnerableDio(String url) async {
  // VULNERABLE: SSRF via Dio
  var dio = Dio();
  return await dio.get(url);
}

// Test 5: URL with user-controlled host
Future<String> vulnerableHost(String host) async {
  // VULNERABLE: User-controlled host
  var client = HttpClient();
  var uri = Uri.parse('http://$host/api/data');
  var request = await client.getUrl(uri);
  var response = await request.close();
  return await response.transform(utf8.decoder).join();
}

// Test 6: URL with user-controlled port
Future<String> vulnerablePort(String port) async {
  // VULNERABLE: User-controlled port (port scanning)
  var client = HttpClient();
  var uri = Uri.parse('http://localhost:$port/health');
  var request = await client.getUrl(uri);
  var response = await request.close();
  return await response.transform(utf8.decoder).join();
}

// Test 7: URL from query parameter
Future<String> vulnerableQueryUrl(Uri requestUri) async {
  // VULNERABLE: URL from query parameter
  var targetUrl = requestUri.queryParameters['url'];
  if (targetUrl != null) {
    var response = await http.get(Uri.parse(targetUrl));
    return response.body;
  }
  return '';
}

// Test 8: Redirect following
Future<String> vulnerableRedirect(String url) async {
  // VULNERABLE: Following redirects to arbitrary locations
  var client = HttpClient();
  client.autoUncompress = true;
  var request = await client.getUrl(Uri.parse(url));
  var response = await request.close();
  return await response.transform(utf8.decoder).join();
}

// Test 9: WebSocket with user URL
Future<WebSocket> vulnerableWebSocket(String wsUrl) async {
  // VULNERABLE: WebSocket SSRF
  return await WebSocket.connect(wsUrl);
}

// Test 10: Socket connection with user host
Future<Socket> vulnerableSocket(String host, int port) async {
  // VULNERABLE: Raw socket SSRF
  return await Socket.connect(host, port);
}

// Test 11: Image/resource fetch
Future<List<int>> vulnerableImageFetch(String imageUrl) async {
  // VULNERABLE: Image URL SSRF
  var response = await http.get(Uri.parse(imageUrl));
  return response.bodyBytes;
}

// Test 12: API proxy pattern
Future<http.Response> vulnerableApiProxy(String endpoint) async {
  // VULNERABLE: API proxy SSRF
  var internalUrl = 'http://internal-api:8080/$endpoint';
  return await http.get(Uri.parse(internalUrl));
}

// Test 13: URL from JSON body
Future<String> vulnerableJsonUrl(String jsonBody) async {
  // VULNERABLE: URL extracted from JSON
  var data = jsonDecode(jsonBody);
  var url = data['callback_url'];
  var response = await http.post(Uri.parse(url), body: 'status=complete');
  return response.body;
}

// Test 14: DNS rebinding potential
Future<String> vulnerableDnsLookup(String hostname) async {
  // VULNERABLE: DNS rebinding
  var addresses = await InternetAddress.lookup(hostname);
  var client = HttpClient();
  var request = await client.getUrl(Uri.parse('http://${addresses.first.address}/'));
  var response = await request.close();
  return await response.transform(utf8.decoder).join();
}

// Test 15: File protocol SSRF
Future<String> vulnerableFileProtocol(String path) async {
  // VULNERABLE: file:// protocol (if supported)
  var response = await http.get(Uri.parse('file://$path'));
  return response.body;
}
